Shopping cart

Someboty opened the pull request.

VirtusX committed:
Transport commit

VirtusX committed:
Transport commit

Someboty committed:
Merge remote-tracking branch 'origin/ShoppingCart' into ShoppingCart

Someboty committed:
Added Transactional to methods, which update data in few steps, separated fields in CreateBookRequestDto for better readability, set book_id and category_id as primary keys in books_categories table

(cherry picked from commit 4deb084b397aa146ddb6641b4f9d5783573c62a4)

Someboty committed:
Changed receiving cart by user_id, fixed style errors

Someboty committed:
Created ShoppingCart and CartItem entities with DTOs, created CartRepository, CartService and CartController for operations with carts and cart items

Zel1oy reviewed (approved):
GJ🔥

andrii-hoienko commented on src/main/java/com/bookshop/dto/cart/CartItemDtoResponse.java:
```diff
@@ -0,0 +1,14 @@
+@Getter
+@Setter
+@EqualsAndHashCode
+public class CartItemDtoResponse {
```
OPTIONAL: It is possible to use @Data instead of 3 annotations

andrii-hoienko commented on src/main/java/com/bookshop/model/Book.java:
```diff
@@ -53,4 +53,12 @@ public class Book {
     @ToString.Exclude
     private Set<Category> categories = new HashSet<>();
+
+    public Book() {
```
It is possible to use NoArgsConstructor annotation

andrii-hoienko commented on src/main/java/com/bookshop/repository/cart/CartRepository.java:
```diff
@@ -0,0 +1,8 @@
+import org.springframework.data.jpa.repository.JpaRepository;
+
+public interface CartRepository extends JpaRepository<ShoppingCart, Long> {
+    ShoppingCart getShoppingCartByUser_Id(Long id);
```
Is it possible to use Java naming convention?

Someboty replied:
That name is generated in interface, I don't think it will work if I try to change it

andrii-hoienko commented on src/main/java/com/bookshop/service/impl/CartServiceImpl.java:
```diff
@@ -0,0 +1,117 @@
+    @Override
+    @Transactional
+    public void clearCart(Long userId) {
+        findCart(userId);
```
What is the purpose of calling this method without usage of value it returned?

Someboty replied:
Checking for user with such userId exists. Without it, repository “deletes” nothing and returns successful code (and user doesn't know, that he didn't delete anything)

andrii-hoienko commented on src/main/java/com/bookshop/service/impl/CartServiceImpl.java:
```diff
@@ -0,0 +1,117 @@
+    @Transactional
+    public void removeCartItem(Long userId, Long cartItemId) {
+        ShoppingCart cart = findCart(userId);
+        findCartItemById(cart, cartItemId);
```
What is the purpose of calling this method without usage of value it returned? If purpose is to check is cartItem exists you should rename method

andrii-hoienko commented on src/main/java/com/bookshop/service/impl/CartServiceImpl.java:
```diff
@@ -0,0 +1,117 @@
+    @Transactional
+    public CartItemDtoResponse createCartItem(Long userId, CreateCartItemDto request) {
+        ShoppingCart cart = findCart(userId);
+        CartItem item = findCartItemByBookId(cart, request.getBookId());
```
I think using optional is preferable way instead of null check

andrii-hoienko commented on src/main/java/com/bookshop/service/impl/CartServiceImpl.java:
```diff
@@ -0,0 +1,117 @@
+        ShoppingCart cart = findCart(userId);
+        if (cart == null) {
+            User user = findUserById(userId);
+            cart = new ShoppingCart();
```
Please move logic of creation and population of models into separate method

andrii-hoienko reviewed (changes requested):
GJ, but some improvements required

Someboty committed:
Replaced Getter, Setter and EqualsAndHashCode by Data annotation
Added NoArgsConstructor annotation
Added examples for some DTOs' values
Removed unnecessary entity calls in CartServiceImpl
Fixed error in setAsRole method

andrii-hoienko reviewed (approved):
GJ
